### PR TITLE
ReAvg/ReVar in rollups violates new unordered destination assumption.

### PR DESCRIPTION
**Pull request by cpwright**

### CODE REVIEW

File: engine/table/src/main/java/io/deephaven/engine/table/impl/by/IntegralChunkedReAvgOperator.java [hunk 1]
```diff
@@ -91,6 +84,12 @@ private void updateResult(ReAvgContext reAvgContext, RowSequence destinationOk,
         for (int ii = 0; ii < size; ++ii) {
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/by/IntegralChunkedReAvgOperator.java:84]: Why do we still need this? I'm pretty sure we just need the new loop.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/by/BigIntegerChunkedReAvgOperator.java [hunk 2]
```diff
@@ -9,12 +9,13 @@
 import io.deephaven.chunk.attributes.Values;
 import io.deephaven.engine.table.ChunkSource;
 import io.deephaven.engine.table.ColumnSource;
+import io.deephaven.engine.table.impl.sort.timsort.LongIntTimsortKernel;
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/by/BigIntegerChunkedReAvgOperator.java:12]: You added unused imports.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/by/DoubleChunkedReAvgOperator.java [hunk 3]
```diff
@@ -12,12 +12,14 @@
 import io.deephaven.chunk.attributes.Values;
 import io.deephaven.engine.table.ChunkSource;
 import io.deephaven.engine.table.ColumnSource;
+import io.deephaven.engine.table.impl.sort.timsort.LongIntTimsortKernel;
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/by/DoubleChunkedReAvgOperator.java:15]: Added unused imports to this file, too.

Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/by/DoubleChunkedReAvgOperator.java:15]: Added unused imports to this file, too.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/by/FloatChunkedReAvgOperator.java [hunk 4]
```diff
@@ -9,12 +9,14 @@
 import io.deephaven.chunk.attributes.Values;
 import io.deephaven.engine.table.ChunkSource;
 import io.deephaven.engine.table.ColumnSource;
+import io.deephaven.engine.table.impl.sort.timsort.LongIntTimsortKernel;
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/by/FloatChunkedReAvgOperator.java:12]: Unused imports here, too.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/by/ReAvgVarOrderingContext.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * For rollups the ReAvg* and ReVar* operators create an ordered keys to read from the result columns from sum
+ * operators. If the slots that we are processing are out-of-order, then we must sort them into order and ensure that
+ * our modified states results properly correspond to the original slot location. In the case that our slots are already
+ * in order, we avoid allocating the sort kernel and statePositions chunk.
+ */
+public class ReAvgVarOrderingContext implements SafeCloseable {
+    public final WritableLongChunk<RowKeys> keyIndices;
+    final int size;
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/by/ReAvgVarOrderingContext.java:23]: Can be private.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/by/ReAvgVarOrderingContext.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    WritableIntChunk<ChunkPositions> statePositions;
+    LongIntTimsortKernel.LongIntSortKernelContext<RowKeys, ChunkPositions> sortKernelContext;
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/by/ReAvgVarOrderingContext.java:25]: Can be private.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/by/ReAvgVarOrderingContext.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import io.deephaven.util.SafeCloseable;
+
+/**
+ * For rollups the ReAvg* and ReVar* operators create an ordered keys to read from the result columns from sum
+ * operators. If the slots that we are processing are out-of-order, then we must sort them into order and ensure that
+ * our modified states results properly correspond to the original slot location. In the case that our slots are already
+ * in order, we avoid allocating the sort kernel and statePositions chunk.
+ */
+public class ReAvgVarOrderingContext implements SafeCloseable {
+    public final WritableLongChunk<RowKeys> keyIndices;
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/by/ReAvgVarOrderingContext.java:22]: Need not be public.

**Review by rcaudy — COMMENTED** [review]
Minor comments/questions, except one operator that needs to be corrected.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/by/DoubleChunkedReAvgOperator.java [hunk 8]
```diff
@@ -60,14 +62,10 @@ public void modifyChunk(BucketedContext context, Chunk<? extends Values> previou
         doBucketedUpdate((ReAvgContext) context, destinations, startPositions, stateModified);
     }
 
+
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/by/DoubleChunkedReAvgOperator.java:65]: The floating point var operator produces a double, but average produces a float.  This one is replicated; the other one is not (it is always FloatChunkedReVarOperator for both float and double).

**Review by rcaudy — DISMISSED** [review]

**Review by rcaudy — APPROVED** [review]

**Pull request merged**